### DISCUSSION TITLE
Texture coordinates with geo map

### DISCUSSION THREAD

**Discussion by shakasaki** [pyvista/pyvista#1859] [Q&A] [answered]
Hello there, I am having some issues understanding how to properly set texture coordinates. I'll give a simple example here but I'm also struggling to understand this when plotting textures (images) onto cylindrical coordinates. But lets start simple.

I have essentially a topography (easting, northing, elevation) and I want to plot maps onto this. All are numpy arrays and when I plot them as numpy arrays the data makes sense (figure 1). When I import these as pyvista textures, the final mesh seems to have shifted some cells in the along the row (Northing) dimension (figure 2). I have tried several options when plotting these with no success. Needless to say I eventually want to plot other textures which are RGB images (figure 3) but these have the same problem. Example code below and data attached. Thanks for the help!

ps. you'll need to unzip the topography file

```py
import numpy as np
import pyvista as pv
import matplotlib.pyplot as plt
from pyvistaqt import BackgroundPlotter

data = np.load('topography.npz')


def create_elevation_surface_from_arrays(elevation: np.array = None, 
                                         east: np.array = None, 
                                         north: np.array = None):
    point_cloud = np.array([east.ravel(), north.ravel(),elevation.ravel()]).T
    cloud = pv.PolyData(point_cloud).extract_geometry().triangulate()
    surface = cloud.delaunay_2d()
    surface.texture_map_to_plane(inplace=True)
    return surface

surf = create_elevation_surface_from_arrays(elevation=data['elevation25m'],
                                            east=data['east25m'],
                                            north=data['north25m'])

# Plot the data as numpy images
fig, ((ax1, ax2, ax3)) = plt.subplots(1,3)
im1 = ax1.imshow(data['elevation25m'])
ax1.set_xlabel('column')
ax1.set_ylabel('row')
ax1.set_title('Elevation (m)')
plt.colorbar(im1, ax=ax1, orientation="horizontal")

im2 = ax2.imshow(data['east25m'])
ax2.set_xlabel('column')
ax2.set_ylabel('row')
ax2.set_title('Easting (m)')
plt.colorbar(im2, ax=ax2, orientation="horizontal")

im3 = ax3.imshow(data['north25m'])
ax3.set_xlabel('column')
ax3.set_ylabel('row')
ax3.set_title('Northing (m)')
plt.colorbar(im3, ax=ax3, orientation="horizontal")



texture = pv.numpy_to_texture(data['elevation25m'])

p = BackgroundPlotter(shape=(1, 2))

p.subplot(0, 0)
p.add_mesh(surf, texture=texture,interpolate_before_map=True)

p.subplot(0, 1)
p.add_mesh(surf, texture=texture,interpolate_before_map=False)
```

![figure1](https://user-images.githubusercontent.com/7264182/143605209-20e6c16d-4080-4fa0-9e7e-1b249a76d1e3.png)
![figure2](https://user-images.githubusercontent.com/7264182/143605213-49e99af2-37f7-4af1-88d2-a3ae11e4186c.png)
![figure3](https://user-images.githubusercontent.com/7264182/143605216-4eb62a7e-c3b5-42e6-b9f1-be3659732426.png)
[topography.zip](https://github.com/pyvista/pyvista/files/7609706/topography.zip)

**Comment by banesullivan** [comment, marked as answer]
I have an example on my website that might be helpful: https://banesullivan.com/pyvista/examples/geological-map.html

Otherwise, give me a little while and I'll try to put something together with your data

**Reply by banesullivan** [reply]
Can you attach the GeoTIFF you're trying to overlay as a texture?

*edit* just noticed you have the numpy array to use as a texture

**Reply by banesullivan** [reply]
Your problem is that you are using the default parameters for `texture_map_to_plane()` which fits a best fitting plane to the data to produce the texture coordinates.

You can see that the texture map has some values outside of the [0, 1] range which is causing the texture to be repeated at the edges.

```py
>> min(surf['Texture Coordinates'][:,0]), max(surf['Texture Coordinates'][:,0])
(-0.046303485, 1.0417274)
```

What you likely want is to simply create the texture coordinates based on the XY coordinates which can be done by:

```py
surf = surf.texture_map_to_plane(use_bounds=True, inplace=False)
# UGH there seems to be a bug with inplace=True here

p = pv.Plotter()
p.add_mesh(surf, texture=texture)
p.show()
```
<img width="536" alt="Screen Shot 2021-11-26 at 6 08 07 PM" src="https://user-images.githubusercontent.com/22067021/143663481-88058476-af2c-407b-b6d9-9eaca13985be.png">

**Reply by shakasaki** [reply]
Hi @banesullivan, your solution works perfectly. I add an example geotif to this surface for completeness of the question. SHaving a look at your example also helps me understand how to set up texture coordinates (I couldn't find a good example of this in the documentation..)

I've been trying to map a borehole image onto a cylinder for a while now, so I'll give it another shot now with some more knowledge on how to set up texture coordinates! Thanks for the help


[swissimage25m_latest.zip](https://github.com/pyvista/pyvista/files/7614522/swissimage25m_latest.zip)
